### ISSUE TITLE
refactor(di): subclasses of BaseException should be immutable

### ISSUE THREAD

**Issue by jeffbcross** [angular/angular#2606]
(Related to #2580 and #2570)

The `AbstractBindingError` class in di/exceptions extends `BaseException` from facade/lang, and provides an `addKey` method to append the error's message with more information. @yjbanov suggests (and I agree) that it would be better to implement the addKey method to return a new `AbstractBindingError`, merging the input with the existing instance's message, to encourage/enforce immutability.

(There may be other subclasses that need to change as well; I just wanted to get this issue open before I left the office).

@vsavkin, I'm happy to make this change if you're not opposed to this (since you're the one who's done the most work on di).

**Comment by jeffbcross** [comment]
(I spoke with @vsavkin in person and he affirmed this approach)

**Comment by mhevery** [comment]
I have tried to merge #2923 in. I have rebased on master, but realized that current implementation has an issue because it turns concrete exception (ie `InstantiationError`) into a (`AbstractBaseError`). To fix this every subclass of `AbstractBindingError` has to implement its own `addKey` method. The issue is that the concrete constructors take `Key` rather then `List<Key>`, so they have to have their signatures changed.

I have decided to abandon this since added code payload and constructor signature complexity does not justify the benefit of the immutability. Given that the exception are only mutated in DI, I see very little value in turning it into immutable code.

Closing.
